issue 2692: missing human teeth when comparing FMA to UBERON

wdduncan opened the pull request:
- assign id ranges to Bill Duncan
- add dosdp-pattern for right teeth

wdduncan commented:
Fixes #2692

wdduncan commented:
@shawntanzk @cmungall @megbalk  
I've added a dosdp-pattern (and data) for creating right human teeth. I'll add left teeth next.

I tested locally by creating a custom ontology to pass into `dosdp-tools generate`. I could not pass in all of UBERON without the process choking. How do you test your dosdp patterns locally?

I tried to use `xrefs` to create an xref on the FMA labels (i.e., exact synonyms) like so:
```
annotations:
  - annotationProperty: "hasExactSynonym"
    value: FMA_label
    xrefs: FMA_xref
```
But I could not get this to work. Is there a way to do this?

Other notes:
- In order to use the `value` key, my variables had to be `data_list_vars`. If I declared the variable as a `data_var`, then I had to substitute the variable using `text: "%s"` like this:
```
- annotations:
  - annotationProperty: "hasExactSynonym"
    text: "%s"
    vars:
       - exact_synonym
```
- I use the `present_in_taxon` annotation to mark these teeth as being present in humans. But, in my output, the value is a string. Should I change to it to an IRI?
- The creation dates are strings. I could not figure out how to get them represented as dataTimes.
- I used `ioi:created_by` with the value of my orcid. I did see some uses of `dc:creator`. Which one is standard/preferred?
- Is there an annotation that is being used for creators/contributors names?

shawntanzk commented:
I think

```
annotations:
  - annotationProperty: "hasExactSynonym"
    value: FMA_label
    xrefs: FMA_xref
```

should be 

```
annotations:
  - annotationProperty: hasExactSynonym
    xrefs: FMA_xref
    text: "%s"
    vars:
      - FMA_label
```

> I use the present_in_taxon annotation to mark these teeth as being present in humans. But, in my output, the value is a string. Should I change to it to an IRI?

yes, I think so

>The creation dates are strings. I could not figure out how to get them represented as dataTimes. 

This is cause your yaml defined date_created as xsd:string I think 

>I used ioi:created_by with the value of my orcid. I did see some uses of dc:creator. Which one is standard/preferred?

Actually dc:contributor as been our go to, but dc:creator works

>Is there an annotation that is being used for creators/contributors names?

dc:contributor 

hope that helps :)

wdduncan commented:
@shawntanzk @megbalk 
> Actually dc:contributor as been our go to, but dc:creator works

I've changed `created_by` to `dc:contributor`.

> This is cause your yaml defined date_created as xsd:string I think

I tried changing the datatype to `xsd:date` and `xsd:dateTime`. The dates are still output as strings.

I tried using this to add an xref to the FMA label:
annotations:
>  annotationProperty: hasExactSynonym
      xrefs: FMA_xref
      text: "%s"
      vars:
        - FMA_label

Unfortunately, this did work either :(
However, the output still does have an xref to the FMA as part of the ontology.

<img width="292" alt="image" src="https://user-images.githubusercontent.com/3186638/202034166-ba5bb560-d938-45d3-95c1-d46bf99b2f2c.png">

Also, the FMA capitalizes the first letter of the labels. Do you want me to lower case it? Or leave as is?

What else needs to be done to get the PR approved?

shawntanzk commented:
@wdduncan fixed some stuff, feel free to take a look at the changes - I honestly usually just find a working dosdp template and work off there, following formats quite strictly, has been the fastest way I've gotten dosdp stuff to work >.<
btws in ODK, the command is `sh run.sh make patterns`
One thing I couldn't get working is the FMA_xref attached to synonyms:
```
annotations:
  - annotationProperty: hasExactSynonym
    xrefs: FMA_xref
    text: "%s"
    vars: 
      - FMA_label
```
DOSDP doesn't seem happy with that xrefs, @hkir-dev any ideas why?
I've removed `xrefs: FMA_xref` for now and the validator seems to be ok when I remove it

shawntanzk commented:
@matentzn - ignore me, I'm stupid, it wasn't in data/default lol

shawntanzk commented:
> Also, the FMA capitalizes the first letter of the labels. Do you want me to lower case it? Or leave as is?

small letters are the convention here, so lower case would be good :) thanks 

> I tried changing the datatype to xsd:date and xsd:dateTime. The dates are still output as strings.

yeah no idea here, @hkir-dev any clues?

hkir-dev commented:
> DOSDP doesn't seem happy with that xrefs, @hkir-dev any ideas why?

According to [dosdp schema](https://github.com/INCATools/dead_simple_owl_design_patterns/blob/master/docs/dosdp_schema.md) xrefs is not supported by annotations. 

We can use `exact_synonym` instead. But it expects data_list_var as value, so this should work: 

```
data_list_vars:              # separate multiple entries by "|"
...
  FMA_label: "xsd:string"        # e.g., Right upper third secondary molar tooth
  FMA_xref: "xsd:string"         # e.g., FMA:55696
  
...

def:
  text: "A %s that is in the right side of the mouth."
  vars:
    - parent_class

exact_synonym:
  value: FMA_label
  xrefs: FMA_xref
```

> I tried changing the datatype to xsd:date and xsd:dateTime. The dates are still output as strings.

Support for datatypes other than string is still an open issue: https://github.com/INCATools/dosdp-tools/issues/136

shawntanzk commented:
@wdduncan think its all fixed now, thanks heaps @hkir-dev :)
let me know if there's anything you aren't sure about the changes

wdduncan commented:
Thanks @shawntanzk !
I'll start working on the left teeth and add it to the PR.

Are the definitions.owl and pattern.owl generated as part of the release, or did you manually add the info to them?
Also, I did know that the tsv was supposed to in `data/default`. You may want to add this info to the documentation.

shawntanzk commented:
> Are the definitions.owl and pattern.owl generated as part of the release, or did you manually add the info to them?

they are, but you can run `sh run.sh make patterns` to generate them, or from the obook tutorial `sh run.sh make ../patterns/definitions.owl -B IMP=false` works too.

> Also, I did know that the tsv was supposed to in data/default. You may want to add this info to the documentation.

yeah maybe not as clear, but it is in the obook tutorial https://oboacademy.github.io/obook/tutorial/dosdp-odk/

wdduncan commented:
> yeah maybe not as clear, but it is in the obook tutorial https://oboacademy.github.io/obook/tutorial/dosdp-odk/

Thanks! I missed that part :(

wdduncan commented:
In the `pattern.owl`, my pattern has the label `xsd:string xsd:string right xsd:string`. Is the something in the `yaml` file that controls the label? If I set it manually, will it be overwritten the next time ontology is rebuild?

wdduncan commented:
Also, how do you handle a case with different values for synonyms and xrefs? 
For example, suppose I. have one set of synonyms that come from `sourc1` and another set of synonyms that come from `source2`. This doesn't work:
```
exact_synonym:
    value: synonym1    
    xrefs: source1

exact_synonym:
    value: synonym2    
    xrefs: source2
```

matentzn commented:
@megbalk this is what I suggested to him on slack - using ROBOT template instead of DOSDP!

shawntanzk commented:
> But, a ROBOT templates may be better in the future, when I need to add more entities to UBERON.

I like the idea of using both, using the DOSDP as you have now, but punting all synonyms (together with any synonyms from any patterns) to a robot template that generates a synonyms component - probably something to implement after next ODK release with better robot template integration :)

wdduncan commented:
@shawntanzk @megbalk I replaced `right_teeth.yaml` with `right_left_teeth.yaml`. I also rebuilt the the `definitions.owl` and `pattern.owl` files.

I think this may be ready.

wdduncan commented:
@megbalk @shawntanzk  
As suggested on the 2022-11-28 UBERON call, I changed the teeth to be `in the righ/left side of` the `upper/lower jaw region`.

There was some discussion about teeth being `attached to` a  jaw region, but this would require more changes. We would need either jaw regions for the upper-right, upper-left, lower-right, lower-left regions, or we would need `attached to right side of`, `attached to left side of` relations. 

I also tried incorporating the use of `oboInOwl:hasSynonymType` to annotate these as human preferred terms, but I couldn't do this as an xref in dosdp. These annotations may need to be added manually.

balhoff commented:
> I also tried incorporating the use of `oboInOwl:hasSynonymType` to annotate these as human preferred terms, but I couldn't do this as an xref in dosdp. These annotations may need to be added manually.

@wdduncan I think you can do this via `printf_annotation` instead of using the shortcut `exact_synonym`.

wdduncan commented:
@balhoff Thanks for the suggestion. Do you have an example of what the syntax would be?

Currently, I use `exact_synonym` so that I can assign an `xref` to the FMA label. E.g.:

<img width="271" alt="image" src="https://user-images.githubusercontent.com/3186638/204661510-a14debbf-3971-42c0-86c1-496997fd7d15.png">

Are you saying that `printf_annotation` would allow me to both annotate the `xref` and annotate the `hasSynonymType` as [preferred term when talking about an instance of this class in Homo sapiens](http://purl.obolibrary.org/obo/uberon/core#HUMAN_PREFERRED)?

wdduncan commented:
Updated the generated ontology to annotate the FMA exact synonym with `HUMAN_PREFERRED`. E.g.:

<img width="452" alt="image" src="https://user-images.githubusercontent.com/3186638/205148601-30da3335-632d-479a-92d9-58784f3d94b4.png">

Thanks @balhoff for your help on this!

For posterity, here is what I did.
1. Added a `human_preferred` column in the tsv with values `UBERON:core#HUMAN_PREFERRED`.
2. Defined a `human_preferred`  var in the `vars` dict.
3. Created the annotations by nesting the xref and synonym type annotations:
```
annotations:
  - annotationProperty: hasExactSynonym
    text: "%s"
    vars:
      - FMA_label
    annotations:
      - annotationProperty: hasDbXref
        value: FMA_xref
      - annotationProperty: hasSynonymType
        var: human_preferred
```

anitacaron commented:
There's a problem with the DOSDP pattern. It does not recognize `some` in the logical axiom `'$parent_class' and ('$lateral_relation' some '$jaw_region')`. That's why the QC is failing.

```bash
2022-12-01 20:08:44.803Z error [] [context: command=terms;pattern=../patterns/dosdp-patterns/right_left_teeth.yaml;input=../patterns/data/default/right_left_teeth.tsv;output=../patterns/data/default/right_left_teeth.txt]
 Failed to parse class expression: '$parent_class' and ('$lateral_relation' some '$jaw_region'):
Encountered some at line 1 column 42. Expected one of:
	or
	and
	)
```

wdduncan commented:
@anitacaron I don't know what to do about this. It runs fine on my machine.  
Do you or @balhoff have any ideas?

balhoff commented:
@wdduncan you found a place that breaks when you try to use a variable for a relation (which is not supported). You probably didn't run the `terms` subcommand locally.

wdduncan commented:
@balhoff Here is the command I'm using:
```
dosdp-tools generate \
    --obo-prefixes=true \
    --infile=../data/default/right_left_teeth.tsv \
    --ontology=temp/teeth.owl \
    --template=right_left_teeth.yaml \
    --outfile=right_left_teeth.owl
```
`right_left_teeth.owl` is an extract of UBERON that I created for efficiency purposes.

wdduncan commented:
What does `dosdp-tools terms` do? `dosdp-tools terms -h` didn't provide much information.  
If I can't dynamically create the relation, then I think I'll have to create separate templates for the right and left teeth :(

wdduncan commented:
@balhoff NVM, I found `terms` on the `dosdp-tools` github page.
```
terms [command options] : dump terms referenced in TSV input and a Dead Simple OWL Design Pattern
      --infile : Input file (TSV or CSV)
```

wdduncan commented:
@balhoff @anitacaron  @shawntanzk @megbalk 
I removed the `right_left_teeth.yaml` and replaced it with two templates: `right_teeth.yaml` and `left_teeth.yaml`.

Can you please review.

anitacaron commented on src/patterns/dosdp-patterns/left_teeth.yaml:
```diff
@@ -0,0 +1,91 @@
+  parent_class: "calcareous tooth"       # UBERON:0001091
+  jaw_region: "jaw region"               # UBERON:0011595
+  FMA_label: "xsd:string"                # e.g., right upper third secondary molar tooth
+  human_preferred: "UBERON:core#HUMAN_PREFERRED"
```
The URI for the synonym type is `http://purl.obolibrary.org/obo/uberon/core#HUMAN_PREFERRED`, so this should be `obo:uberon/core#HUMAN_PREFERRED`.
Please, also change the tsv file.

wdduncan replied:
Thanks @anitacaron. This has been changed.

anitacaron replied:
Can you change it in the two YAML files, please?

wdduncan replied:
done

anitacaron commented on src/patterns/dosdp-patterns/left_teeth.yaml:
```diff
@@ -0,0 +1,91 @@
+
+data_list_vars:                # separate multiple entries by "|"
+  synonym: "xsd:string"        # e.g., ADA tooth 1, FDI tooth 18
+  creator: "xsd:string"        # e.g., https://orcid.org/0000-0001-9625-1899
```
The type creator is not string because it's an ORCID.

wdduncan replied:
Thank @anitacaron. I changed the range of `creator` to `xsd:anyURI`.

anitacaron commented on src/patterns/dosdp-patterns/left_teeth.yaml:
```diff
@@ -0,0 +1,91 @@
+annotationProperties:
+  hasExactSynonym: "oboInOwl:hasExactSynonym"
+  present_in_taxon: "RO:0002175"
+  contributor: "dc:contributor"
```
Same for right_teeth yml file

wdduncan replied:
I change the pattern_iri.

wdduncan replied:
dct:contributor is now used instead of dc:contributor.

wdduncan commented on src/patterns/dosdp-patterns/left_teeth.yaml:
```diff
@@ -0,0 +1,91 @@
+pattern_name: right__left_teeth
+pattern_iri: http://purl.obolibrary.org/obo/uberon/patterns/right_left_teeth.yaml
```
I change the pattern_iri.

wdduncan commented:
@anitacaron I pushed a `definitions.owl` file with the classes created from dosdp-templates. Do you want me to merge this into the edit file? I don't know what the process is for merging in the new classes.

anitacaron commented on src/patterns/dosdp-patterns/left_teeth.yaml:
```diff
@@ -0,0 +1,91 @@
+vars:
+  parent_class: "calcareous tooth"       # UBERON:0001091
+  jaw_region: "jaw region"               # UBERON:0011595
+  FMA_label: "xsd:string"                # e.g., right upper third secondary molar tooth
```
Move `FMA_label` to data_vars. The `vars` key is to place OWL classes.

wdduncan replied:
done
```
data_list_vars:                # separate multiple entries by "|"
  synonym: "xsd:string"        # e.g., ADA tooth 1, FDI tooth 18
  date_created: "xsd:dateTime" # e.g., 2022-11-13T00:00:00Z; note: this shows up as a string in Protege
  FMA_xref: "xsd:string"       # e.g., FMA:55696
```

anitacaron commented on src/patterns/dosdp-patterns/left_teeth.yaml:
```diff
@@ -0,0 +1,91 @@
+
+data_list_vars:                # separate multiple entries by "|"
+  synonym: "xsd:string"        # e.g., ADA tooth 1, FDI tooth 18
+  creator: "xsd:anyURI"        # e.g., https://orcid.org/0000-0001-9625-1899
+  date_created: "xsd:dateTime" # e.g., 2022-11-13T00:00:00Z; note: this shows up as a string in Protege
+  taxon: "xsd:string"          # e.g., obo:NCBITaxon_9606; note: the full IRI is being used as a value
+  FMA_xref: "xsd:string"       # e.g., FMA:55696
```
Move `taxon` to `vars` because it's an OWL class. Change it to `taxon: "Homo sapiens"`, as you defined before. Also move `creator` to `vars` and add value "https://orcid.org/0000-0001-9625-1899" to be an entity in the OWL.
Move `date_created` and `FMA_xref` to `data_vars` as they are a single value.

anitacaron replied:
Unfortunately, it's not possible to have other type than `xsd:string`, that's why date_created is generated as string even though you defined as `xsd:dateTime`. I'll create an issue to fix this using SPARQL.

wdduncan replied:
done
```
vars:
  parent_class: "calcareous tooth"  # UBERON:0001091
  jaw_region: "jaw region"          # UBERON:0011595
  taxon: "Homo sapiens"             # obo:NCBITaxon_9606
  creator: "xsd:anyURI"             # e.g., https://orcid.org/0000-0001-9625-1899
  human_preferred: "obo:uberon/core#HUMAN_PREFERRED"
```

```
annotationProperty: present_in_taxon
    var: taxon
```
```
annotationProperty: contributor
    var: creator
```

anitacaron replied:
Sorry you didn't make the second change I suggested.

> Move date_created and FMA_xref to data_vars as they are a single value.

anitacaron replied:
And also, this part

> [move creator to vars] add value "https://orcid.org/0000-0001-9625-1899" to be an entity in the OWL.

wdduncan replied:
@anitacaron 
I changed the requested `data_list_vars` to `data_vars`.

> [move creator to vars] add value "https://orcid.org/0000-0001-9625-1899" to be an entity in the OWL.

This has been done. See [here](https://github.com/obophenotype/uberon/blob/15a13a958719695f1b42cf9430bd0efb62bee37a/src/patterns/dosdp-patterns/right_teeth.yaml#L34). 
```
vars:
  parent_class: "calcareous tooth"  # UBERON:0001091
  jaw_region: "jaw region"          # UBERON:0011595
  taxon: "Homo sapiens"             # obo:NCBITaxon_9606
  creator: "xsd:anyURI"             # e.g., https://orcid.org/0000-0001-9625-1899
  human_preferred: "obo:uberon/core#HUMAN_PREFERRED"
```
Since the range is `xsd:anyURI` a URI is output in the `dcterms:contributor` annotation.

anitacaron replied:
I was thinking of adding your ORCID as the default value.

anitacaron commented on src/patterns/dosdp-patterns/left_teeth.yaml:
```diff
@@ -0,0 +1,91 @@
+
+annotations:
+  - annotationProperty: present_in_taxon
+    value: taxon
```
change `value` to `var` so it will be generated as an entity in OWL

wdduncan replied:
done

anitacaron commented on src/patterns/dosdp-patterns/left_teeth.yaml:
```diff
@@ -0,0 +1,91 @@
+  - annotationProperty: hasExactSynonym
+    value: synonym
+  - annotationProperty: contributor
+    value: creator
```
change `value` to `var` so it will be generated as an entity in OWL

wdduncan replied:
done

anitacaron reviewed (changes requested):
After updating the `left_teeth.yaml` and `right_teeth.yaml`, please run `sh run.sh make patterns` to update the `definitions.owl` file.

anitacaron commented:
#gogoeditdiff

shawntanzk commented:
@megbalk would you mind reviewing this please? I think its alright, esp given its pulled over from FMA, but just want a 2nd eye and I know like nothing about teeth stuff. Thanks :)

anitacaron commented:
Thank you, @wdduncan, for all changes! My review of the DOS-DP is done, but the ontology part also needs a review.

megbalk commented:
will do! How do I check the ontology? Like do you mean the reasoning?

megbalk commented on src/patterns/dosdp-patterns/left_teeth.yaml:
```diff
@@ -0,0 +1,95 @@
+pattern_name: right__left_teeth
```
change to left_teeth

megbalk commented on src/patterns/dosdp-patterns/left_teeth.yaml:
```diff
@@ -0,0 +1,95 @@
+
+relations:
+  in lateral side of: "BSPO:0000126"
+  in right side of: "BSPO:0000121"
```
remove this

wdduncan replied:
Remove the relations?

megbalk replied:
in right side of since this is for the left teeth

megbalk commented on src/patterns/dosdp-patterns/left_teeth.yaml:
```diff
@@ -0,0 +1,95 @@
+  text: "%s %s"   # e.g, right upper third secondary molar tooth
+  vars:
+    - laterality   # e.g., right
+    - parent_class # e.g., right upper third secondary molar tooth
```
remove "right"

wdduncan replied:
Sorry ... I'm not following. We need laterality to distinguish teeth. Right?

megbalk replied:
we do, but the parent class should just be "upper third secondary molar tooth"

megbalk commented on src/patterns/dosdp-patterns/left_teeth.yaml:
```diff
@@ -0,0 +1,95 @@
+def:
+  text: "A %s that is in the right side of the %s."
+  # e.g., A upper third secondary molar tooth that is in the right side of the upper jaw region.
+  vars:
```
can you add laterality as the second vars and then have def be: A %s that is %s the %s?

megbalk commented on src/patterns/dosdp-patterns/left_teeth.yaml:
```diff
@@ -0,0 +1,95 @@
+    - parent_class # e.g., right upper third secondary molar tooth
+
+def:
+  text: "A %s that is in the right side of the %s."
```
this should be left, not right

megbalk commented on src/patterns/dosdp-patterns/left_teeth.yaml:
```diff
@@ -0,0 +1,95 @@
+  synonym: "xsd:string"        # e.g., ADA tooth 1, FDI tooth 18
+  
+name:
+  text: "%s %s"   # e.g, right upper third secondary molar tooth
```
change to left, not right

wdduncan replied:
Sure ... that is just a an example in a comment though.

megbalk commented on src/patterns/dosdp-patterns/right_teeth.yaml:
```diff
@@ -0,0 +1,95 @@
+pattern_name: right__left_teeth
```
change to right_teeth

megbalk commented on src/patterns/dosdp-patterns/right_teeth.yaml:
```diff
@@ -0,0 +1,95 @@
+relations:
+  in lateral side of: "BSPO:0000126"
+  in right side of: "BSPO:0000121"
+  in left side of: "BSPO:0000120"
```
remove left side

wdduncan replied:
`in left side of` isn't harming anything ...

megbalk commented on src/patterns/dosdp-patterns/right_teeth.yaml:
```diff
@@ -0,0 +1,95 @@
+  text: "%s %s"   # e.g, right upper third secondary molar tooth
+  vars:
+    - laterality   # e.g., right
+    - parent_class # e.g., right upper third secondary molar tooth
```
remove laterality

wdduncan replied:
laterality isn't harming anything ...

wdduncan replied:
done

megbalk reviewed (changes requested):
see comments

wdduncan commented:
@megbalk  
I made a number of the changes you requested.  
I'm not understanding your comments about laterality. Don't we need that in the label? Is there another labeling pattern you want to use?

anitacaron commented:
@megbalk here you can check the diff https://github.com/obophenotype/uberon/pull/2713#issuecomment-1378550134

megbalk commented:
Looks good! I am not familiar with FMA or ADA naming, so I am relying on @wdduncan to check that they are correct.

I also mentioned that the order of teeth naming may be different, but that shouldn't affect the pattern.

wdduncan commented:
The current naming pattern dynamically prepends the laterality (“left”/“right”) to the name of the parent class. 
```
name:
  text: "%s %s"   # e.g, left upper third secondary molar tooth
  vars:
    - laterality   # e.g., left
    - parent_class # e.g., left upper third secondary molar tooth
 ```
For example, this data in `left_teeth.tsv`:
| parent_class | laterality |
| ---- | --- |
| UBERON:0018577 | left |

yields the label `left upper third secondary molar tooth` b/c the parent class' label is `upper third secondary molar tooth`.

Do you want keep the laterality first? If so, then I can recreate the left/right labels after the parent classes tooth labels have been changed. E.g. If the parent label is `secondary upper third tooth` the left tooth will be `left secondary upper third tooth`. Otherwise, I will have implement a more complicated pattern for the labels.

wdduncan commented:
@megbalk comments changed to avoid confusion ;)
